File: Presentation/Windows/PermissionsPanelController.swift
import Cocoa

/// Protocol for permissions panel callbacks
protocol PermissionsPanelDelegate: AnyObject {
    var dataConsentKey: String { get }
}

/// Controller for the permissions setup panel
class PermissionsPanelController {
    private weak var delegate: PermissionsPanelDelegate?

    // UI Elements
    private(set) var panel: NSView?
    private var screenRecordingStatusLabel: NSTextField?
    private var dataConsentStatusLabel: NSTextField?
    private var permissionCheckTimer: Timer?

    init(delegate: PermissionsPanelDelegate) {
        self.delegate = delegate
    }

    deinit {
        stopMonitoring()
    }

    /// Setup the permissions panel in the parent view
    func setup(in parentView: NSView) {
        // Center the panel vertically and horizontally
        let panelWidth: CGFloat = 560
        let panelHeight: CGFloat = 360
        let panel = NSVisualEffectView(frame: NSRect(
            x: (parentView.frame.width - panelWidth) / 2,
            y: (parentView.frame.height - panelHeight) / 2,
            width: panelWidth,
            height: panelHeight
        ))
        panel.autoresizingMask = [.minXMargin, .maxXMargin, .minYMargin, .maxYMargin]
        panel.blendingMode = .withinWindow
        panel.material = .hudWindow
        panel.state = .active
        panel.alphaValue = 0.96
        panel.wantsLayer = true
        panel.layer?.cornerRadius = 20
        panel.layer?.borderWidth = 2.0
        panel.layer?.borderColor = NSColor.appleGold.withAlphaComponent(0.6).cgColor
        parentView.addSubview(panel)
        self.panel = panel

        // Title
        let title = NSTextField(labelWithString: "🔐 Setup Required")
        title.frame = NSRect(x: 40, y: panel.frame.height - 60, width: panel.frame.width - 80, height: 35)
        title.font = .systemFont(ofSize: 26, weight: .bold)
        title.textColor = .white
        title.alignment = .center
        panel.addSubview(title)

        // Subtitle
        let subtitle = NSTextField(labelWithString: "Complete the following steps to use Interview Master")
        subtitle.frame = NSRect(x: 40, y: panel.frame.height - 88, width: panel.frame.width - 80, height: 20)
        subtitle.font = .systemFont(ofSize: 13, weight: .regular)
        subtitle.textColor = NSColor.white.withAlphaComponent(0.75)
        subtitle.alignment = .center
        panel.addSubview(subtitle)

        // Screen Recording Permission (top row)
        createPermissionRow(
            in: panel,
            yOffset: panel.frame.height - 175,
            icon: "📸",
            title: "Screen Recording",
            description: "Capture coding problems during interviews",
            buttonTitle: "Open Settings",
            action: #selector(openScreenRecordingSettings),
            isScreenRecording: true
        )

        // Anthropic Data Consent (bottom row)
        createPermissionRow(
            in: panel,
            yOffset: panel.frame.height - 260,
            icon: "🤖",
            title: "AI Data Sharing",
            description: "Send screenshots to Anthropic Claude for analysis",
            buttonTitle: "I Consent",
            action: #selector(grantDataConsent),
            isScreenRecording: false
        )

        // Hint text at bottom
        let hint = NSTextField(labelWithString: "Global shortcuts (⌘B, ⌘S, ⌘Enter) work from any app")
        hint.frame = NSRect(x: 40, y: 20, width: panel.frame.width - 80, height: 16)
        hint.font = .systemFont(ofSize: 11, weight: .regular)
        hint.textColor = NSColor.white.withAlphaComponent(0.5)
        hint.alignment = .center
        panel.addSubview(hint)

        // Check permissions and update UI
        updateStatus()
        startMonitoring()
    }

    private func createPermissionRow(in panel: NSView, yOffset: CGFloat, icon: String, title: String, description: String, buttonTitle: String, action: Selector, isScreenRecording: Bool) {
        // Container
        let container = NSView(frame: NSRect(x: 40, y: yOffset, width: panel.frame.width - 80, height: 70))
        panel.addSubview(container)

        // Icon
        let iconLabel = NSTextField(labelWithString: icon)
        iconLabel.frame = NSRect(x: 0, y: 20, width: 40, height: 30)
        iconLabel.font = .systemFont(ofSize: 28)
        iconLabel.alignment = .center
        container.addSubview(iconLabel)

        // Title & Description
        let titleLabel = NSTextField(labelWithString: title)
        titleLabel.frame = NSRect(x: 50, y: 38, width: 200, height: 20)
        titleLabel.font = .systemFont(ofSize: 15, weight: .semibold)
        titleLabel.textColor = .white
        container.addSubview(titleLabel)

        let descLabel = NSTextField(labelWithString: description)
        descLabel.frame = NSRect(x: 50, y: 18, width: 300, height: 18)
        descLabel.font = .systemFont(ofSize: 12, weight: .regular)
        descLabel.textColor = NSColor.white
        container.addSubview(descLabel)

        // Status indicator
        let statusLabel = NSTextField(labelWithString: "⚠️ Required")
        statusLabel.frame = NSRect(x: container.frame.width - 180, y: 42, width: 120, height: 18)
        statusLabel.font = .systemFont(ofSize: 12, weight: .medium)
        statusLabel.textColor = .systemOrange
        statusLabel.alignment = .right
        container.addSubview(statusLabel)

        // Store reference based on type
        if isScreenRecording {
            screenRecordingStatusLabel = statusLabel
        } else {
            dataConsentStatusLabel = statusLabel
        }

        // Action button
        let button = NSButton(frame: NSRect(x: container.frame.width - 120, y: 12, width: 120, height: 28))
        button.title = buttonTitle
        button.bezelStyle = .rounded
        button.target = self
        button.action = action
        button.font = .systemFont(ofSize: 12, weight: .medium)
        container.addSubview(button)
    }

    @objc func openScreenRecordingSettings() {
        NSWorkspace.shared.open(URL(string: "x-apple.systempreferences:com.apple.preference.security?Privacy_ScreenCapture")!)
    }

    @objc func grantDataConsent() {
        guard let key = delegate?.dataConsentKey else { return }
        UserDefaults.standard.set(true, forKey: key)
        updateStatus()
    }

    func updateStatus() {
        guard let key = delegate?.dataConsentKey else { return }

        let hasScreenRecording = CGPreflightScreenCaptureAccess()
        let hasDataConsent = UserDefaults.standard.bool(forKey: key)

        screenRecordingStatusLabel?.stringValue = hasScreenRecording ? "✅ Enabled" : "⚠️ Required"
        screenRecordingStatusLabel?.textColor = hasScreenRecording ? .appleGreen : .systemOrange

        dataConsentStatusLabel?.stringValue = hasDataConsent ? "✅ Granted" : "⚠️ Required"
        dataConsentStatusLabel?.textColor = hasDataConsent ? .appleGreen : .systemOrange

        // Hide panel if both permissions granted
        if hasScreenRecording && hasDataConsent {
            panel?.isHidden = true
            stopMonitoring()
        } else {
            panel?.isHidden = false
        }
    }

    func startMonitoring() {
        permissionCheckTimer = Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { [weak self] _ in
            self?.updateStatus()
        }
    }

    func stopMonitoring() {
        permissionCheckTimer?.invalidate()
        permissionCheckTimer = nil
    }
}

File: Presentation/Windows/PermissionsPanelController.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Media;
using System.Windows.Threading;

namespace InterviewMasterApp.Presentation.Windows
{
    /// <summary>
    /// Delegate for permissions panel callbacks.
    /// </summary>
    public interface IPermissionsPanelDelegate
    {
        string DataConsentKey { get; }
    }

    /// <summary>
    /// Permissions panel controller (WPF). Ported from PermissionsPanelController.swift.
    /// </summary>
    public class PermissionsPanelController
    {
        private readonly IPermissionsPanelDelegate _delegate;
        private Border? _panel;
        private TextBlock? _accessibilityStatus;
        private TextBlock? _screenRecordingStatus;
        private TextBlock? _dataConsentStatus;
        private DispatcherTimer? _timer;

        public PermissionsPanelController(IPermissionsPanelDelegate @delegate)
        {
            _delegate = @delegate;
        }

        public void Setup(Panel parent)
        {
            var panel = new Border
            {
                Width = 560,
                Height = 440,
                Background = new SolidColorBrush(Color.FromArgb(245, 25, 25, 30)),
                CornerRadius = new CornerRadius(20),
                BorderBrush = Brushes.Gold,
                BorderThickness = new Thickness(2),
                HorizontalAlignment = HorizontalAlignment.Center,
                VerticalAlignment = VerticalAlignment.Center,
                Padding = new Thickness(20)
            };

            var stack = new StackPanel { Orientation = Orientation.Vertical, VerticalAlignment = VerticalAlignment.Center };
            panel.Child = stack;

            var title = new TextBlock { Text = "Setup Required", Foreground = Brushes.White, FontSize = 24, FontWeight = FontWeights.Bold, HorizontalAlignment = HorizontalAlignment.Center };
            var subtitle = new TextBlock { Text = "Complete the following steps to use Interview Master", Foreground = Brushes.LightGray, FontSize = 12, HorizontalAlignment = HorizontalAlignment.Center, Margin = new Thickness(0, 4, 0, 12) };
            stack.Children.Add(title);
            stack.Children.Add(subtitle);

            CreateRow(stack, "Accessibility", "Global hotkeys from any app", "Open Settings", out _accessibilityStatus, OpenAccessibilitySettings);
            CreateRow(stack, "Screen Recording", "Capture coding problems", "Open Settings", out _screenRecordingStatus, OpenScreenRecordingSettings);
            CreateRow(stack, "AI Data Sharing", "Send screenshots for analysis", "I Consent", out _dataConsentStatus, GrantDataConsent);

            var hint = new TextBlock { Text = "Shortcuts work from any app", Foreground = Brushes.Gray, FontSize = 11, HorizontalAlignment = HorizontalAlignment.Center, Margin = new Thickness(0, 12, 0, 0) };
            stack.Children.Add(hint);

            parent.Children.Add(panel);
            _panel = panel;

            UpdateStatus();
            StartMonitoring();
        }

        private void CreateRow(Panel parent, string title, string description, string buttonTitle, out TextBlock statusLabel, Action clickHandler)
        {
            var row = new Grid { Margin = new Thickness(0, 8, 0, 8) };
            row.ColumnDefinitions.Add(new ColumnDefinition { Width = new GridLength(1, GridUnitType.Star) });
            row.ColumnDefinitions.Add(new ColumnDefinition { Width = new GridLength(120) });

            var stack = new StackPanel { Orientation = Orientation.Vertical };
            stack.Children.Add(new TextBlock { Text = title, Foreground = Brushes.White, FontSize = 14, FontWeight = FontWeights.SemiBold });
            stack.Children.Add(new TextBlock { Text = description, Foreground = Brushes.LightGray, FontSize = 11 });

            statusLabel = new TextBlock { Text = "Required", Foreground = Brushes.Orange, FontSize = 11, HorizontalAlignment = HorizontalAlignment.Right };

            var button = new Button { Content = buttonTitle, Width = 110, Margin = new Thickness(0, 4, 0, 0) };
            button.Click += (_, _) => clickHandler();

            Grid.SetColumn(stack, 0);
            Grid.SetColumn(statusLabel, 1);
            Grid.SetColumn(button, 1);

            row.Children.Add(stack);
            row.Children.Add(statusLabel);
            row.Children.Add(button);
            parent.Children.Add(row);
        }

        private void OpenAccessibilitySettings()
        {
            // Best-effort: open Windows privacy settings
            try { System.Diagnostics.Process.Start(new System.Diagnostics.ProcessStartInfo("ms-settings:privacy-accessibility") { UseShellExecute = true }); } catch { }
        }

        private void OpenScreenRecordingSettings()
        {
            try { System.Diagnostics.Process.Start(new System.Diagnostics.ProcessStartInfo("ms-settings:privacy-screenrecording") { UseShellExecute = true }); } catch { }
        }

        private void GrantDataConsent()
        {
            var key = _delegate.DataConsentKey;
            LocalSettings.SetBool(key, true);
            UpdateStatus();
        }

        public void UpdateStatus()
        {
            var hasAccessibility = CheckAccessibility();
            var hasScreenRecording = CheckScreenRecording();
            var hasDataConsent = LocalSettings.GetBool(_delegate.DataConsentKey);

            SetStatus(_accessibilityStatus, hasAccessibility);
            SetStatus(_screenRecordingStatus, hasScreenRecording);
            SetStatus(_dataConsentStatus, hasDataConsent);

            if (_panel != null)
            {
                _panel.Visibility = (hasAccessibility && hasScreenRecording && hasDataConsent) ? Visibility.Collapsed : Visibility.Visible;
            }
        }

        public void StartMonitoring()
        {
            _timer = new DispatcherTimer { Interval = TimeSpan.FromSeconds(1) };
            _timer.Tick += (_, _) => UpdateStatus();
            _timer.Start();
        }

        public void StopMonitoring()
        {
            _timer?.Stop();
            _timer = null;
        }

        private static void SetStatus(TextBlock? label, bool enabled)
        {
            if (label == null) return;
            label.Text = enabled ? "Enabled" : "Required";
            label.Foreground = enabled ? Brushes.LightGreen : Brushes.Orange;
        }

        // Windows does not have direct equivalents for these macOS permissions.
        private static bool CheckAccessibility() => true;
        private static bool CheckScreenRecording() => true;
    }

    internal static class LocalSettings
    {
        private static readonly string FilePath = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".interview-master-settings.json");

        public static bool GetBool(string key)
        {
            var data = ReadAll();
            return data.TryGetValue(key, out var val) && val;
        }

        public static void SetBool(string key, bool value)
        {
            var data = ReadAll();
            data[key] = value;
            WriteAll(data);
        }

        private static Dictionary<string, bool> ReadAll()
        {
            try
            {
                if (!File.Exists(FilePath)) return new Dictionary<string, bool>();
                var json = File.ReadAllText(FilePath);
                return JsonSerializer.Deserialize<Dictionary<string, bool>>(json) ?? new Dictionary<string, bool>();
            }
            catch
            {
                return new Dictionary<string, bool>();
            }
        }

        private static void WriteAll(Dictionary<string, bool> data)
        {
            try
            {
                var json = JsonSerializer.Serialize(data, new JsonSerializerOptions { WriteIndented = true });
                File.WriteAllText(FilePath, json);
            }
            catch
            {
                // ignore persistence errors
            }
        }
    }
}

